compiler/src/compiler/optimize/constant_folding.rs

```diff
@@ -0,0 +1,289 @@
+    pub fn fold_constants(&mut self) {
+        self.body
+            .visit_with_visible(&mut |_, expression, visible, _| {
+                let Expression::Call {
+                function,
+                arguments,
+                responsible,
+            } = expression else { return; };
+                let Expression::Builtin(builtin) = visible.get(*function) else { return; };
+
+                if let Some(result) = Self::run_builtin(*builtin, arguments, *responsible, visible)
+                {
+                    let evaluated_call = match result {
+                        Ok(return_value) => return_value,
+                        Err(panic_reason) => {
+                            let mut body = Body::new();
+                            let reason = body.push_with_new_id(
+                                &mut self.id_generator,
+                                Expression::Text(panic_reason),
+                            );
+                            body.push_with_new_id(
+                                &mut self.id_generator,
+                                Expression::Panic {
+                                    reason,
+                                    responsible: *responsible,
+                                },
+                            );
+                            Expression::Multiple(body)
+                        }
+                    };
+                    *expression = evaluated_call;
+                }
+            });
+    }
+
+    /// This function tries to run a builtin, requiring a minimal amount of
+    /// static knowledge. For example, it can find out that the result of
+    /// `builtinEquals $3 $3` is `True`, even if the value of `$3` is not known
+    /// at compile-time.
+    ///
+    /// Returns `None` if the call couldn't be evaluated statically. Returns
+    /// `Some(Ok(expression))` if the call successfully completed with a return
+    /// value. Returns `Some(Err(reason))` if the call panics.
+    fn run_builtin(
+        builtin: BuiltinFunction,
+        arguments: &[Id],
+        responsible: Id,
+        visible: &VisibleExpressions,
+    ) -> Option<Result<Expression, String>> {
+        // warn!("Constant folding candidate: builtin{builtin:?}");
+        // warn!(
+        //     "Arguments: {}",
+        //     arguments.iter().map(|arg| format!("{arg}")).join(", ")
+        // );
+        // warn!(
+        //     "Expressions:\n{}",
+        //     expressions
+        //         .iter()
+        //         .map(|(id, expr)| format!("{id}: {expr}"))
+        //         .join("\n")
+        // );
+
+        Some(Ok(match builtin {
+            BuiltinFunction::Equals => {
+                if arguments.len() != 2 {
+                    return Some(Err("wrong number of arguments".to_string()));
+                }
+
+                let a = arguments[0];
+                let b = arguments[1];
+
+                let are_equal = a == b || a.semantically_equals(b, visible)?;
```

Comment:
The primitive equality check is included in `semantically_equals(…)`

```suggestion
                let are_equal = a.semantically_equals(b, visible)?;
```